Installation issue: `gcc@14.2` on macos

### Steps to reproduce the issue

gcc@14.2 was [just released ](https://gcc.gnu.org/pipermail/gcc/2024-August/244512.html) but does not support darwin: `*** Configuration aarch64-apple-darwin23.5.0 not supported`. Homebrew is still on 14.1 so there are no available patches.
I'm not familiar with where the patches are derived. I suspect this work https://github.com/iains/gcc-14-branch ?


```console
$ spack spec -I gcc@14.2
Input spec
--------------------------------
 -   gcc@14.2

Concretized
--------------------------------
 -   gcc@14.2.0%apple-clang@15.0.0~binutils~bootstrap~graphite~mold~nvptx~piclibs~strip build_system=autotools build_type=RelWithDebInfo languages='c,c++,fortran' arch=darwin-sonoma-m1
[+]      ^diffutils@3.10%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]          ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gawk@5.3.0%apple-clang@15.0.0~nls build_system=autotools arch=darwin-sonoma-m1
[+]          ^libsigsegv@2.14%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]          ^readline@8.2%apple-clang@15.0.0 build_system=autotools patches=bbf97f1 arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gmp@6.3.0%apple-clang@15.0.0+cxx build_system=autotools libs=shared,static arch=darwin-sonoma-m1
[+]          ^autoconf@2.72%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]          ^automake@1.16.5%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]          ^libtool@2.4.7%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]              ^findutils@4.9.0%apple-clang@15.0.0 build_system=autotools patches=440b954 arch=darwin-sonoma-m1
[+]          ^m4@1.4.19%apple-clang@15.0.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1
[+]      ^libiconv@1.17%apple-clang@15.0.0 build_system=autotools libs=shared,static arch=darwin-sonoma-m1
[+]      ^libtool@2.4.7%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]          ^findutils@4.9.0%apple-clang@15.0.0 build_system=autotools patches=440b954 arch=darwin-sonoma-m1
[+]          ^m4@1.4.19%apple-clang@15.0.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=darwin-sonoma-m1
[+]              ^diffutils@3.9%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]                  ^libiconv@1.17%apple-clang@15.0.0 build_system=autotools libs=shared,static arch=darwin-sonoma-m1
[+]              ^libsigsegv@2.14%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]      ^mpc@1.3.1%apple-clang@15.0.0 build_system=autotools libs=shared,static arch=darwin-sonoma-m1
[+]      ^mpfr@4.2.1%apple-clang@15.0.0 build_system=autotools libs=shared,static arch=darwin-sonoma-m1
[+]          ^autoconf-archive@2023.02.20%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]      ^perl@5.40.0%apple-clang@15.0.0+cpanm+opcode+open+shared+threads build_system=generic arch=darwin-sonoma-m1
[+]          ^berkeley-db@18.1.40%apple-clang@15.0.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=darwin-sonoma-m1
[+]          ^bzip2@1.0.8%apple-clang@15.0.0~debug~pic+shared build_system=generic arch=darwin-sonoma-m1
[+]          ^gdbm@1.23%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]      ^texinfo@7.1%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]          ^gettext@0.22.5%apple-clang@15.0.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=darwin-sonoma-m1
[+]              ^libxml2@2.10.3%apple-clang@15.0.0+pic~python+shared build_system=autotools arch=darwin-sonoma-m1
[+]              ^tar@1.34%apple-clang@15.0.0 build_system=autotools zip=pigz arch=darwin-sonoma-m1
[+]                  ^pigz@2.8%apple-clang@15.0.0 build_system=makefile arch=darwin-sonoma-m1
[+]              ^xz@5.4.6%apple-clang@15.0.0~pic build_system=autotools libs=shared,static arch=darwin-sonoma-m1
[+]          ^ncurses@6.5%apple-clang@15.0.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=darwin-sonoma-m1
[+]              ^pkgconf@2.2.0%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
[+]      ^zlib@1.3.1%apple-clang@15.0.0+optimize+pic+shared build_system=makefile arch=darwin-sonoma-m1
[+]      ^zstd@1.5.6%apple-clang@15.0.0~ipo+programs build_system=cmake build_type=Release compression=none generator=make libs=shared,static arch=darwin-sonoma-m1
[+]          ^cmake@3.30.2%apple-clang@15.0.0~doc+ncurses+ownlibs build_system=generic build_type=Release patches=dbc3892 arch=darwin-sonoma-m1
[+]              ^curl@8.7.1%apple-clang@15.0.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=secure_transport arch=darwin-sonoma-m1
[+]                  ^nghttp2@1.62.0%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m1
```


### Error message

<details><summary>Error message</summary>

<pre>
==> Installing gcc-14.2.0-ki2xxo7f42n63obualtwsrfz6dli7tr3 [99/115]
==> No binary for gcc-14.2.0-ki2xxo7f42n63obualtwsrfz6dli7tr3 found: installing from source
==> Fetching https://ftpmirror.gnu.org/gcc/gcc-14.2.0/gcc-14.2.0.tar.xz
==> Ran patch() for gcc
==> gcc: Executing phase: 'autoreconf'
==> gcc: Executing phase: 'configure'
==> gcc: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j10' 'V=1'

2 errors found in build log:
     3634    checking for AF_INET6... yes
     3635    checking for _LK_LOCK... no
     3636    checking for off_t... no
     3637    checking if mkdir takes one argument... no
     3638    yes
     3639    checking for size_t... *** Configuration aarch64-apple-darwin23.5.0 not supported
  >> 3640    make[1]: *** [Makefile:4637: configure-gcc] Error 1
     3641    make[1]: *** Waiting for unfinished jobs....
     3642    yes
     3643    checking for ssize_t... yes
     3644    checking for uintptr_t... yes
     3645    checking for ptrdiff_t... yes
     3646    checking for uint64_t... yes

     ...

     3714    config.status: creating Makefile
     3715    config.status: creating config.h
     3716    config.status: executing depdir commands
     3717    mkdir .deps
     3718    config.status: executing default-1 commands
     3719    make[1]: Leaving directory '/Users/cmarsh/Documents/science/code/spack/stage/spack-stage-gcc-14.2.0-ki2xxo7f42n63obualtwsrfz6dli7tr3/spack-src/spack-build'
  >> 3720    make: *** [Makefile:1059: all] Error 2

</pre></details>


### Information on your system

spack debug report
* **Spack:** 0.23.0.dev0 (9364250ad8b2c60e1fd00ce06b81c2c3a5c0d632)
* **Python:** 3.12.4
* **Platform:** darwin-sonoma-m1
* **Concretizer:** clingo

### Additional information

@alalazo  @michaelkuhn

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -439,6 +439,11 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage, CompilerPackage):
             sha256="1529cff128792fe197ede301a81b02036c8168cb0338df21e4bc7aafe755305a",
             when="@14.1.0 target=aarch64:",
         )
+        patch(
+            "https://raw.githubusercontent.com/Homebrew/formula-patches/f30c309442a60cfb926e780eae5d70571f8ab2cb/gcc/gcc-14.2.0-r2.diff",
+            sha256="6c0a4708f35ccf2275e6401197a491e3ad77f9f0f9ef5761860768fa6da14d3d",
+            when="@14.2.0 target=aarch64:",
+        )
         conflicts("+bootstrap", when="@11.3.0,13.1: target=aarch64:")
 
         # Use -headerpad_max_install_names in the build,

